presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java

```diff
@@ -378,9 +407,39 @@ public void unassignShard(long tableId, UUID shardUuid, String nodeIdentifier)
                     })
                     .build();
         }
-        catch (DBIException e) {
-            throw metadataError(e);
+    }
+
+    @Override
+    public long beginTransaction()
+    {
+        return dao.insertTransaction();
+    }
+
+    @Override
+    public void rollbackTransaction(long transactionId)
+    {
+        dao.finalizeTransaction(transactionId, false);
+    }
+
+    private static void commitTransaction(ShardManagerDao dao, long transactionId)
+    {
+        if (dao.finalizeTransaction(transactionId, true) != 1) {
+            throw new PrestoException(TRANSACTION_CONFLICT, "Transaction commit failed. Please retry the operation.");
         }
+        dao.deleteCreatedShards(transactionId);
+        dao.deleteCreatedShardNodes(transactionId);
+    }
+
+    @Override
+    public void recordCreatedShard(long transactionId, UUID shardUuid, String nodeIdentifier)
+    {
+        int nodeId = getOrCreateNodeId(nodeIdentifier);
+        runTransaction(dbi, (handle, status) -> {
+            ShardManagerDao dao = handle.attach(ShardManagerDao.class);
+            dao.insertCreatedShard(shardUuid, transactionId);
+            dao.insertCreatedShardNode(shardUuid, nodeId, transactionId);
```

Comment:
You are missing the shard assignments through `assignShard` in `RaptorSplitManager` and  `ShardEjector`. 
We should be recording those in ``created_shard_node`` as well.